tests/drivers/i2s/i2s_api/src/test_i2s_states.c

```diff
@@ -351,6 +351,12 @@ void test_i2s_state_stopping_neg(void)
 
 	ret = rx_block_read(dev_i2s_rx, 0);
 	zassert_equal(ret, TC_PASS, NULL);
+
+	/* This is incase the RX channel is stuck in STOPPING state.
+	 * Clear out the state before running the next test.
+	 */
+	ret = i2s_trigger(dev_i2s_rx, I2S_DIR_RX, I2S_TRIGGER_DROP);
+	zassert_equal(ret, 0, "RX DROP trigger failed");
```

Comment:
At this point we've sent two data blocks and we read two data blocks. The driver should not be any more in the stopping state. Seems like the driver implementation is the problem.

Reply:
This in the case where the STOP trigger is called after the data has already been received and the dma_rx callback has finished executing.

Reply:
I am referring to the STOP trigger that is sent on line 339. I understand the test is verifying return of error conditions and in my case it is passing those parts. However at the end of the test the RX port was left in STOPPING state as the data was already received I feel prior to calling the STOP trigger. Hence I add the lines of code so we can run the next test.

Reply:
OK, I got it. You are right. This needs to be fixed. The testcase relies on the fact that the transmission on the I2S interface is slow. It shouldn't. I propose to fix it differently though. Please remove the DROP trigger that you added and instead move the lines
```
	/* All but one data block read, stop reception */
	ret = i2s_trigger(dev_i2s_rx, I2S_DIR_RX, I2S_TRIGGER_STOP);
	zassert_equal(ret, 0, "RX STOP trigger failed");
```
to just after the following section
```
	/* All data written, flush TX queue and stop the transmission */
	ret = i2s_trigger(dev_i2s_tx, I2S_DIR_TX, I2S_TRIGGER_DRAIN);
	zassert_equal(ret, 0, "TX DRAIN trigger failed");
```
So we will guarantee that the STOP trigger on the RX interface is sent just after DRAIN trigger on the TX interface.

Reply:
I tried the moving the lines as you suggested. The test now fails in the below lines as the I2S RX port is in the READY state after going to STOPPING and receiving the data packets.

```
	/* Send invalid triggers, expect failure */
	ret = i2s_trigger(dev_i2s_rx, I2S_DIR_RX, I2S_TRIGGER_START);
	zassert_equal(ret, -EIO, NULL);

```

Reply:
Oh, well. It looks like the multiple calls to `i2s_trigger`, `zassert_equal` take too long on your SoC. By the time the testcase arrives at the section testing RX failure triggers the RX is already done with reception.

I propose to put one more data block in the TX queue, so the testcase will have more time to run the checks. I've added one `tx_block_write` and one accompanying `rx_block_read`. I propose to write the testcase as follows.

```
void test_i2s_state_stopping_neg(void)
{
	const struct device *dev_i2s;
	int ret;

	dev_i2s = device_get_binding(I2S_DEV_NAME);
	zassert_not_null(dev_i2s, "device " I2S_DEV_NAME " not found");

	/* Prefill TX queue */
	ret = tx_block_write(dev_i2s, 0, 0);
	zassert_equal(ret, TC_PASS, NULL);
	ret = tx_block_write(dev_i2s, 0, 0);
	zassert_equal(ret, TC_PASS, NULL);

	/* Start reception */
	ret = i2s_trigger(dev_i2s, I2S_DIR_RX, I2S_TRIGGER_START);
	zassert_equal(ret, 0, "RX START trigger failed");

	/* Start transmission */
	ret = i2s_trigger(dev_i2s, I2S_DIR_TX, I2S_TRIGGER_START);
	zassert_equal(ret, 0, "TX START trigger failed");

	ret = tx_block_write(dev_i2s, 0, 0);
	zassert_equal(ret, TC_PASS, NULL);

	ret = rx_block_read(dev_i2s, 0);
	zassert_equal(ret, TC_PASS, NULL);

	/* All data written, flush TX queue and stop the transmission */
	ret = i2s_trigger(dev_i2s, I2S_DIR_TX, I2S_TRIGGER_DRAIN);
	zassert_equal(ret, 0, "TX DRAIN trigger failed");

	/* Send invalid triggers, expect failure */
	ret = i2s_trigger(dev_i2s, I2S_DIR_TX, I2S_TRIGGER_START);
	zassert_equal(ret, -EIO, NULL);
	ret = i2s_trigger(dev_i2s, I2S_DIR_TX, I2S_TRIGGER_STOP);
	zassert_equal(ret, -EIO, NULL);
	ret = i2s_trigger(dev_i2s, I2S_DIR_TX, I2S_TRIGGER_DRAIN);
	zassert_equal(ret, -EIO, NULL);
	ret = i2s_trigger(dev_i2s, I2S_DIR_TX, I2S_TRIGGER_PREPARE);
	zassert_equal(ret, -EIO, NULL);

	ret = rx_block_read(dev_i2s, 0);
	zassert_equal(ret, TC_PASS, NULL);

	/* All but one data block read, stop reception */
	ret = i2s_trigger(dev_i2s, I2S_DIR_RX, I2S_TRIGGER_STOP);
	zassert_equal(ret, 0, "RX STOP trigger failed");

	/* Send invalid triggers, expect failure */
	ret = i2s_trigger(dev_i2s, I2S_DIR_RX, I2S_TRIGGER_START);
	zassert_equal(ret, -EIO, NULL);
	ret = i2s_trigger(dev_i2s, I2S_DIR_RX, I2S_TRIGGER_STOP);
	zassert_equal(ret, -EIO, NULL);
	ret = i2s_trigger(dev_i2s, I2S_DIR_RX, I2S_TRIGGER_DRAIN);
	zassert_equal(ret, -EIO, NULL);
	ret = i2s_trigger(dev_i2s, I2S_DIR_RX, I2S_TRIGGER_PREPARE);
	zassert_equal(ret, -EIO, NULL);

	ret = rx_block_read(dev_i2s, 0);
	zassert_equal(ret, TC_PASS, NULL);
}
```

I have verified that the code works on `sam_e70_xplained` board with `i2s_sam_ssc.c` driver.

Reply:
This does not work for me. I start seeing I2S RX overflows. Moving the below 2 lines right below the trigger to DRAIN the TX fixes the overflow, however in that case I still see the original issue.
```
	ret = rx_block_read(dev_i2s, 0);
	zassert_equal(ret, TC_PASS, NULL);
```

Reply:
This the RX FIFO overflowing at the hardware level triggering an interrupt in the driver.

Reply:
But this shouldn't happen. I'm asking to make sure it's not a bug in the driver. Unfortunately, I don't have a MIMXRT685-EVK board to test the PR locally.

Reply:
@mknp. I have updated the I2S receive implementation to add double buffering. I am not able to pass the test with your changes. However I still see the issue where the RX channel is stuck in STOPPING state. I will add another commit to update the I2S RX DMA implementation.

I will also add a check to not allow transfers below 16-bit till we are able to add a new API.

Reply:
What do you mean by double buffering? I don't think we need to change RX implementation, unless you found a bug.

> I will also add a check to not allow transfers below 16-bit till we are able to add a new API.

Yes, that's probably the best course of action. Let's disable transfers below 16-bits for now. We can add it later when we figure out how to support it in API.

Reply:
After recent modifications to the driver, do the new version of `test_i2s_state_stopping_neg` testcase that I proposed in the previous comment still fail? Is the failure connected with RX FIFO over- or underflow?